mongoengine/context_managers.py

```diff
@@ -230,3 +230,11 @@ def set_write_concern(collection, write_concerns):
     combined_concerns = dict(list(collection.write_concern.document.items()))
     combined_concerns.update(write_concerns)
     yield collection.with_options(write_concern=WriteConcern(**combined_concerns))
+
+
+@contextmanager
+def set_read_write_concern(collection, write_concerns, read_concern):
+    combined_write_concerns = dict(collection.write_concern.document.items())
+    combined_write_concerns.update(write_concerns)
+
+    yield collection.with_options(write_concern=WriteConcern(**combined_write_concerns), read_concern=read_concern)
```

Comment:
Warning: Both of these context managers are very misleading in that... They're not really context managers. They're simple functions which return a PyMongo collection with specific options. They do not change any context.

```
In [14]: from pymongo.read_concern import ReadConcern

In [15]: from mongoengine.context_managers import set_read_write_concern

In [16]: from mongoengine import *

In [17]: connect("testdb")
Out[17]: MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True, read_preference=Primary())

In [18]: class User(Document):
    ...:     name = StringField()
    ...:

In [19]: User.objects
Out[19]: []

In [20]: with set_read_write_concern(User._collection, {"w": 0, "j": False}, ReadConcern(level='local')):
    ...:     # You might be tempted to think that this MongoEngine operation is going to
    ...:     # use whatever write/read concern you've set in the context manager above.
    ...:     # It most definitely will not. That's kind of a blessing in disguise though,
    ...:     # because using such a *working* context manager would not be thread-safe
    ...:     # and would probably cause issues even when using coroutines / gevent.
    ...:     qs = User.objects.filter(name="aaa")
    ...:     list(qs)
    ...:     print(qs._collection.write_concern.document)
    ...:     print(qs._cursor.collection.write_concern.document)
    ...:     print(qs._collection.read_concern.document)
    ...:     print(qs._cursor.collection.read_concern.document)
    ...:
{}
{}
{}
{}
```

Should we rewrite them as regular functions?

Reply:
As this is a cherry-pick [from upstream](https://github.com/MongoEngine/mongoengine/commit/8e17e42e26d944d90bdd16bfacb8d07879768055) I think we should prefer to keep it as close to upstream as possible. This way we can at least _partially_ benefit from upstream documentation and changes, and makes any cherry-picks in the future easier. Don't you agree @wojcikstefan ?

Reply:
Yea, I suppose I agree. I hope we can fix the upstream, too (though that's a breaking change, so would need a new major version bump).

Reply:
Side note: I'm also concerned about the interface of this context manager where `read_concern` is passed as an instance of `ReadConcern`, but `write_concern` is a plain dict.

Reply:
I'm not really sure I wholeheartedly agree with your interpretation about context managers though - I don't think it's correct to make assumptions about what context the context managers are managing other than the resource they return.

In this case the context manager is returning a collections object with some options applied so if you use that object you're fine. I think it's incorrect to assume what it should or should not modify outside of that - I don't see myself expecting that your example of `Users.objects...` should have the options applied at all.

Reply:
Why is it a context manager then, if a) it doesn't change any context within the code block it's executed in, b) it doesn't perform any setup cleanup on exit? This is just a function with a return value.

Reply:
Yep - I agree there's no _need_ for it to be a context manager and could just as well be a function. But in the end it's functionally equivalent and is just a design choice.